Softeq.XToolkit.PushNotifications/PushTokenSynchronizer.cs

```diff
@@ -0,0 +1,153 @@
+        private readonly string _isTokenSavedOnServerKey = $"{nameof(IPushTokenSynchronizer)}_token_saved_on_server_key";
+
+        private readonly IPushTokenStorageService _pushTokenStorageService;
+        private readonly IRemotePushNotificationsService _remotePushNotificationsService;
+        private readonly IInternalSettings _internalSettings;
+        private readonly IPushNotificationsHandler _pushNotificationsHandler;
+        private readonly ILogger _logger;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="PushTokenSynchronizer"/> class.
+        /// </summary>
+        /// <param name="pushNotificationsHandler">Handles received push notifications events.</param>
+        /// <param name="internalSettings">Parses remote messages into common format.</param>
+        /// <param name="pushTokenStorageService">Stores push notification token and it's status.</param>
+        /// <param name="remotePushNotificationsService">Propagates push notification token to remote server.</param>
+        /// <param name="logManager">Provides logging.</param>
+        public PushTokenSynchronizer(
+            IPushNotificationsHandler pushNotificationsHandler,
+            IInternalSettings internalSettings,
+            IPushTokenStorageService pushTokenStorageService,
+            IRemotePushNotificationsService remotePushNotificationsService,
+            ILogManager logManager)
+        {
+            _pushNotificationsHandler = pushNotificationsHandler;
+            _internalSettings = internalSettings;
+            _pushTokenStorageService = pushTokenStorageService;
+            _remotePushNotificationsService = remotePushNotificationsService;
+
+            _logger = logManager.GetLogger<PushTokenSynchronizer>();
+        }
+
+        protected virtual TimeSpan TokenSendRetryDelay { get; } = TimeSpan.FromSeconds(10);
+
+        protected virtual int MaxAttemptsCount { get; } = int.MaxValue;
+
+        public bool IsTokenSavedOnServer
+        {
+            get => _internalSettings.GetValueOrDefault(_isTokenSavedOnServerKey, default(bool));
+            set => _internalSettings.AddOrUpdateValue(_isTokenSavedOnServerKey, value);
+        }
+
+        private bool IsTokenRegisteredInSystem
+        {
+            get => _internalSettings.GetValueOrDefault(_isTokenRegisteredInSystemKey, default(bool));
+            set => _internalSettings.AddOrUpdateValue(_isTokenRegisteredInSystemKey, value);
+        }
+
+        /// <inheritdoc />
+        public Task ResendTokenToServerIfNeedAsync()
+        {
+            if (IsTokenRegisteredInSystem && !IsTokenSavedOnServer)
+            {
+                return DoSendTokenToServer(_pushTokenStorageService.PushToken);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public async Task UnregisterFromRemotePushNotificationsAsync()
```

Comment:
Seems like we need to change `IsRegisteredInSystem` in this case

Reply:
Fixed